added solution code to lizard era: beginning (#2962)
cpinitiative/usaco-guide: solutions/cf-585D.mdx

---
id: cf-585D
source: CF
title: "Lizard Era: Beginning"
author: Albert Ye
---

[Official Analysis](https://codeforces.com/blog/entry/20898)

> We notice that the size of the array is 25. A naive brute-force would take $\mathcal O(3^n)$ which is too slow because $n = 25$ and $3^25 = 8.47 \cdot 10^10 > 5 \cdot 10^9$. But, a meet-in-the-middle solution would take at most $\mathcal O(\sqrt{3^n} \log (\sqrt{3^n}))$, which is a significant optimization. 

Reviewer: Prove the time complexity after the MIM-optimization... What makes it so fast? What did you optimize?

## Implementation

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

const int MAXN = 25;

int n, mid;

// the added / removed approval from each hero for each action
// hero 0 = Lynn
// hero 1 = Meliana
// hero 2 = Worrigan

> // henceforth referred to as L,M,W

Reviewer 2: multiline comments like:
```
/*
 * one
 * two
 * three
 */
```

	rec1(i+1, app_L, app_M + d[i][1], app_W + d[i][2], opt * 3 + 2); // 2 - take M, W
}

const int INF = 1e9+7;
int maxs=-INF, lopt=0, ropt=0; // max sum and the optimal decision paths

// recursion for the second half
void rec2(int i, int app_L, int app_M, int app_W, int opt) {
	if(i < mid) {
		// want to make up for the difference offsets calculated in the first half
		if(m.find({app_M - app_L, app_W - app_L}) != m.end()) {
			if(m[{app_M - app_L, app_W - app_L}].first + app_L > maxs) {
				maxs = m[{app_M - app_L, app_W - app_L}].first + app_L;
				lopt = m[{app_M - app_L, app_W - app_L}].second;
				ropt = opt;
			}
		}
		return;
	}
	rec2(i-1, app_L + d[i][0], app_M + d[i][1], app_W, opt * 3);
	rec2(i-1, app_L + d[i][0], app_M, app_W + d[i][2], opt * 3 + 1);
	rec2(i-1, app_L, app_M + d[i][1], app_W + d[i][2], opt * 3 + 2);
}

string to[3] = {"LM", "LW", "MW"}; // same order as desc'd in rec1
int main() {
	cin >> n;
	for(int i = 0; i < n; i++) {
		cin >> d[i][0] >> d[i][1] >> d[i][2];
	}
	mid = n/2;
	// 0 -> mid
	rec1(0, 0, 0, 0, 0);
	// n-1 -> mid
	rec2(n-1, 0, 0, 0, 0);
	if(maxs > -INF) {
		// retrace steps

> 		vector<int> r1, r2;

Reviewer: I'd rather you use better variable names than so many comments like this. Comments in specific places are fine, but people dont know what general variables such as `m`, `r1`, `r2`, `ropt` means...

---
id: cf-585D
source: CF
title: Lizard Era: Beginning
author: Albert Ye
---

[Official Analysis](https://codeforces.com/blog/entry/20898)

> In sum, we notice that the size of the array is 25. A naive brute-force would take $\mathcal O(3^n)$ which is too slow. But, a meet-in-the-middle solution would take at most $\mathcal O(\sqrt{3^n} \log (\sqrt{3^n}))$, which is a significant optimization. 

Reviewer: ```suggestion
We notice that the size of the array is 25. A naive brute-force would take $\mathcal O(3^n)$ which is too slow. But, a meet-in-the-middle solution would take at most $\mathcal O(\sqrt{3^n} \log (\sqrt{3^n}))$, which is a significant optimization. 
```
?

> In sum, we notice that the size of the array is 25. A naive brute-force would take $\mathcal O(3^n)$ which is too slow. But, a meet-in-the-middle solution would take at most $\mathcal O(\sqrt{3^n} \log (\sqrt{3^n}))$, which is a significant optimization. 

Reviewer: Also why for the second sentence? prove it!

Reviewer 3: agreed!

> ## Implementation

Reviewer 2: time complexity?

<LanguageSection>
<CPPSection>
```cpp
#include <bits/stdc++.h>

using namespace std;

const int MAXN = 25;

int n, mid;
// the added / removed approval from each hero for each action
int d[MAXN][3];
map<pair<int, int>, pair<int, int> > m;

// each recursion has the index we're at, the approvals we have per hero, and the steps we took to get here

> void rec1(int i, int d0, int d1, int d2, int opt)

Reviewer: oh god these variable names.... please make them more informative rather than slapping on a comment

{
	if(i >= mid)
	{
		// find the largest happiness you can get for the first half
		if(m.count({d0-d1, d0-d2}))
		{
			if(m[{d0-d1, d0-d2}].first < d0) m[{d0-d1, d0-d2}] = {d0, opt};
		} else 
		{

> 			m[{d0-d1, d0-d2}] = {d0, opt};

Reviewer 2: space out operators

		}
		return;

> 	}

Reviewer 2: recommended to keep braces on same line:
```
for () {

}
```

Author: oh will fix this part

	rec1(i+1, d0+d[i][0], d1+d[i][1], d2, opt*3);
	rec1(i+1, d0+d[i][0], d1, d2+d[i][2], opt*3+1);
	rec1(i+1, d0, d1+d[i][1], d2+d[i][2], opt*3+2);
}

const int INF = 1e9+7;
int maxs=-INF, lopt=0, ropt=0; // max sum and the optimal decision paths

> void rec2(int i, int d0, int d1, int d2, int opt)

Reviewer: what does rec1 and rec2 do????

{
	if(i < mid)
	{
		if(m.find({d1-d0, d2-d0}) != m.end())
		{
			if(m[{d1-d0, d2-d0}].first + d0 > maxs)
			{
				maxs = m[{d1-d0, d2-d0}].first + d0;
				lopt = m[{d1-d0, d2-d0}].second;
				ropt = opt;
			}
		}
		return;
	}
	rec2(i-1, d0+d[i][0], d1+d[i][1], d2, opt*3);
	rec2(i-1, d0+d[i][0], d1, d2+d[i][2], opt*3+1);
	rec2(i-1, d0, d1+d[i][1], d2+d[i][2], opt*3+2);
}

string to[3] = {"LM", "LW", "MW"};
int main()
{
	cin >> n;
	for(int i = 0; i < n; i++)
	{
		cin >> d[i][0] >> d[i][1] >> d[i][2];
	}
	mid = n/2;
	// 0 -> mid
	rec1(0, 0, 0, 0, 0);
	// n-1 -> mid
	rec2(n-1, 0, 0, 0, 0);
	if(maxs > -INF)
	{
		// retrace steps
		vector<int> r1, r2;

> 		for(int i = 0; i < n/2; i++)

Reviewer: inconsistent spacing

---
id: cf-585D
source: CF
title: "Lizard Era: Beginning"
author: Albert Ye
---

[Official Analysis](https://codeforces.com/blog/entry/20898)

We notice that the size of the array is 25. A naive brute-force would take at least $\mathcal O(3^n)$ which is too slow

> because $n = 25$ and $3^25 = 8.47 \cdot 10^10 > 5 \cdot 10^9$. 

Reviewer 4: ```suggestion
because $n = 25$ and $3^25 = 8.47 \cdot 10^10 > 5 \cdot 10^9$. 
```

I would recommend using https://aops.com/texer to render your latex first to see how it comes out before pasting it here. Time for some heavy change reqs!

Author: ok my bad !!!!!!!!!

> But, a meet-in-the-middle solution would take at most $\mathcal O(\sqrt{3^n} \log (\sqrt{3^n}))$,
> which is a significant optimization. Let $a_l, a_m, a_w$ be the approvals of the three heros.

Reviewer 3: ```suggestion
However, a meet-in-the-middle solution would be much faster, taking $\mathcal O(\sqrt{3^n} \log (\sqrt{3^n}))$.
Let $a_l, a_m, a_w$ be the approvals of the three heros.
```

Reviewer 3: also the transition is really weird here

We run a brute-force solution on the left half of the array and store the differences $a_l-a_m, a_l-a_w$
as the keys of a map. These differences point to the approval value of one of the hero (let's just say $a_l$
for simplicity) and the current decision path.

We then run a similar brute force over the right half of the array. Except this time, we try to make the final 
differences $a_l-a_m, a_l-a_w$ equal 0. This means that if $a_l-a_m = p$ and $a_l-a_w = q$ in the left half, then
$a_l-a_m = -p$ and $a_l-a_w = -q$ must be true in the right half. For each pair of differences $(a_l-a_m, a_l-a_w)$
in the right half, we try to find an equivalent pair $(a_m-a_l, a_w-a_l)$ in the left half.

If we find such a pair, we check the total approval $a_l$, which is guaranteed to be equal to $a_m, a_w$ if we 
perform the steps given in the left half and then the steps given in the right half. We try to maximize $a_l$.

Since we run $\mathcal O(3^k \log 3^k)$ brute force on each half, we have $k = \frac{n}{2}$. We add a log factor to this
because the `map` data structure adds $\mathcal O(\log n)$ time. Therefore, the total complexity is 

> $\mathcal O(3^{\frac n2}) = \mathcal O(\sqrt{3^n} \log \sqrt{3^n})$, which fits the boundaries.

Author: bruv no one finna see this

## Implementation

<LanguageSection>
<CPPSection>

```cpp
#include <bits/stdc++.h>
using namespace std;

const int MAXN = 25;

int n, mid;

/* the added / removed approval from each hero for each action
 * hero 0 = Lynn
 * hero 1 = Meliana
 * hero 2 = Worrigan
 * henceforth referred to as L,M,W

>  */

Reviewer 3: ```suggestion
/*
 * the added / removed approval from each hero for each action
 * hero 0 = Lynn
 * hero 1 = Meliana
 * hero 2 = Worrigan
 * henceforth referred to as L,M,W
 */

int d[MAXN][3];

> map<pair<int, int>, pair<int, int> > m;

Reviewer 3: ```suggestion
map<pair<int, int>, pair<int, int>> m;
```

// in each recursive function, we store i, the approval of heroes L, M, W, and then an integer saving our decision path

> // recur

Reviewer 3: what is `recur`

void rec1(int i, int app_L, int app_M, int app_W, int path) {

> 	if(i >= mid) {

Reviewer 3: ```suggestion
	if (i >= mid) {
```

		// find all arrangements of ppl in the first half
		if(m.count({app_L - app_M, app_L - app_W})) {
			if(m[{app_L - app_M, app_L - app_W}].first < app_L) m[{app_L - app_M, app_L - app_W}] = {app_L, path};
		} else {
			m[{app_L - app_M, app_L - app_W}] = {app_L, path};
		}
		return;
	}

> 	rec1(i+1, app_L + d[i][0], app_M + d[i][1], app_W, path * 3); // 0 - take L, M

Reviewer 3: spaces

	rec1(i+1, app_L + d[i][0], app_M, app_W + d[i][2], path * 3 + 1); // 1 - take L, W
	rec1(i+1, app_L, app_M + d[i][1], app_W + d[i][2], path * 3 + 2); // 2 - take M, W
}

> const int INF = 1e9+7;

Reviewer 3: or just,,,, `INT32_MAX`...

int maxs=-INF, left_path=0, right_path=0; // max sum and the optimal decision paths for each half

// recursion for the second half
void rec2(int i, int app_L, int app_M, int app_W, int path) {
	if(i < mid) {
		// want to make up for the difference offsets calculated in the first half
		if(m.find({app_M - app_L, app_W - app_L}) != m.end()) {
			if(m[{app_M - app_L, app_W - app_L}].first + app_L > maxs) {
				maxs = m[{app_M - app_L, app_W - app_L}].first + app_L;
				left_path = m[{app_M - app_L, app_W - app_L}].second;
				right_path = path;
			}
		}
		return;
	}
	rec2(i-1, app_L + d[i][0], app_M + d[i][1], app_W, path * 3);
	rec2(i-1, app_L + d[i][0], app_M, app_W + d[i][2], path * 3 + 1);
	rec2(i-1, app_L, app_M + d[i][1], app_W + d[i][2], path * 3 + 2);
}

> string to[3] = {"LM", "LW", "MW"}; // same order as desc'd in rec1

Reviewer 4: desc'd? that could mean descended, described, etc etc.

int main() {
	cin >> n;
	for(int i = 0; i < n; i++) {
		cin >> d[i][0] >> d[i][1] >> d[i][2];
	}
	mid = n/2;
	// 0 -> mid
	rec1(0, 0, 0, 0, 0);
	// n-1 -> mid
	rec2(n-1, 0, 0, 0, 0);

> 	if(maxs > -INF) {

Reviewer 4: ```suggestion
	if (maxs > -INF) {
```
